firmware/src/core/utils/Utils.cpp

```diff
@@ -211,3 +211,83 @@ uint16_t Utils::rgb565dim(uint16_t color, uint8_t brightness, bool swapBytes) {
     }
     return result;
 }
+
+void Utils::rgb565dimBitmap(uint16_t *pixel565, size_t length, uint8_t brightness, bool swapBytes) {
+    for (int i = 0; i < length; i++) {
+        pixel565[i] = rgb565dim(pixel565[i], brightness, swapBytes);
+    }
+}
+
+// Function to extract R, G, B from a 16-bit RGB565 pixel
+uint32_t Utils::rgb565ToRgb888(uint16_t rgb565, bool swapBytes) {
+    uint8_t r, g, b;
+
+    if (swapBytes) {
+        // Swap bytes
+        rgb565 = (rgb565 >> 8) | (rgb565 << 8);
+    }
+    r = (rgb565 >> 11) & 0x1F; // Extract red (5 bits)
+    g = (rgb565 >> 5) & 0x3F; // Extract green (6 bits)
+    b = rgb565 & 0x1F; // Extract blue (5 bits)
+
+    // Scale to 8 bits
+    r = (r * 255) / 31;
+    g = (g * 255) / 63;
+    b = (b * 255) / 31;
+
+    // Pack R, G, B into a single 32-bit color
+    return (r << 16) | (g << 8) | b;
+}
+
+// Function to scale RGB888 color to RGB565 format
+uint16_t Utils::rgb888ToRgb565(uint32_t rgb888, bool swapBytes) {
+    uint8_t r = (rgb888 >> 16) & 0xFF;
+    uint8_t g = (rgb888 >> 8) & 0xFF;
+    uint8_t b = rgb888 & 0xFF;
+
+    uint16_t pixel = ((r >> 3) << 11) | ((g >> 2) << 5) | (b >> 3);
+
+    if (swapBytes) {
+        // Swap bytes
+        pixel = (pixel >> 8) | (pixel << 8);
+    }
+    return pixel;
+}
+
+// Function to apply grayscale and map to target color
+uint16_t Utils::grayscaleToTargetColor(uint8_t grayscale, uint32_t targetColor565, float brightness, bool swapBytes) {
+    uint32_t targetColor888 = rgb565ToRgb888(targetColor565, false);
```

Comment:
Thanks, that's right. I moved the calculations for targetColor888 and its components to colorizeImageData, so that we only need to calculate it once per call.